### PR TITLE
drivers: can: stm32: Remove STM CAN_Init

### PR DESCRIPTION
**Pull request by alexanderwachter**
Removed the STM32 CAN_Init function and implemented the initialization
in the driver.

This commit resolves the dependency on STM32 HAL and the legacy HAL API.

### CODE REVIEW

File: drivers/can/can_stm32.c [hunk 1]
```diff
@@ -856,7 +918,8 @@ static void config_can_1_irq(CAN_TypeDef *can)
 		    can_stm32_tx_isr, DEVICE_GET(can_stm32_1), 0);
 	irq_enable(DT_CAN_1_IRQ_SCE);
 #endif
-	can->IER |= CAN_IT_TME | CAN_IT_ERR | CAN_IT_FMP0 | CAN_IT_FMP1;
+	can->IER |= CAN_IER_TMEIE | CAN_IER_ERRIE | CAN_IER_FMPIE0 |
```
Comment (erwango) [inline drivers/can/can_stm32.c]: Is that stritcly HAL related?
If not, can you move it to a separate commit?

File: drivers/can/can_stm32.c [hunk 2]
```diff
@@ -360,15 +422,15 @@ int can_stm32_send(struct device *dev, const struct zcan_frame *msg,
 
 	if (transmit_status_register & CAN_TSR_TME0) {
 		LOG_DBG("Using mailbox 0");
-		mailbox = &can->sTxMailBox[CAN_TXMAILBOX_0];
+		mailbox = &can->sTxMailBox[0];
```
Comment (erwango) [inline drivers/can/can_stm32.c]: Maybe we can keep the #define's (or similar ones) and avoid magic numbers

File: drivers/can/can_stm32.c [hunk 3]
```diff
@@ -305,13 +347,33 @@ static int can_stm32_init(struct device *dev)
...
 	if (ret) {
 		return ret;
 	}
 
+	/* Leave sleep mode after reset*/
+	can->MCR &= ~CAN_MCR_SLEEP;
+
 	cfg->config_irq(can);
-	can->IER |= CAN_IT_TME;
+	can->IER |= CAN_IER_TMEIE;
```
Comment (erwango) [inline drivers/can/can_stm32.c]: Is that HAL removal related?

File: drivers/can/Kconfig.stm32 [hunk 4]
```diff
@@ -9,7 +9,6 @@
 config CAN_STM32
 	bool "STM32 CAN Driver"
 	depends on SOC_FAMILY_STM32
-	select USE_STM32_HAL_CAN
 	help
 	  Enable STM32 CAN Driver (tested on stm32F0 series)
```
Comment (erwango) [inline drivers/can/Kconfig.stm32]: Is that still right ?

File: drivers/can/can_stm32.c [hunk 5]
```diff
@@ -81,6 +83,10 @@ void can_stm32_rx_isr_handler(CAN_TypeDef *can, struct can_stm32_data *data)
 		/* Release message */
 		can->RF0R |= CAN_RF0R_RFOM0;
 	}
+
+	if (can->RF0R & CAN_RF0R_FOVR0) {
```
Comment (erwango) [inline drivers/can/can_stm32.c]: Is that HAL related?

**Review by erwango — CHANGES_REQUESTED** [review]
I'm globally fine with the change, but not CAN proficient. We'd need other to review and test.
On which series have you been able to test? 
Seems to me that some changes are not strictly relaetd to HAL removal, so would be clearer to split in more commits if this is the case.

**Comment by erwango** [comment]
@lowlander, you might be interested

File: drivers/can/can_stm32.c (hunk 1 above)
Comment (alexanderwachter) [reply drivers/can/can_stm32.c]: It is. CAN_IT_TME is from HAL and CAN_IER_TMEIE from mcu headers.

File: drivers/can/can_stm32.c (hunk 2 above)
Comment (alexanderwachter) [reply drivers/can/can_stm32.c]: Same here. CAN_TXMAILBOX_* are definitions from HAL

File: drivers/can/can_stm32.c (hunk 3 above)
Comment (alexanderwachter) [reply drivers/can/can_stm32.c]: Same here

File: drivers/can/Kconfig.stm32 (hunk 4 above)
Comment (alexanderwachter) [reply drivers/can/Kconfig.stm32]: Yes. All dependencies from HAL are removed.

File: drivers/can/can_stm32.c (hunk 5 above)
Comment (alexanderwachter) [reply drivers/can/can_stm32.c]: No. I can shift this o an extra commit.

**Comment by alexanderwachter** [comment]
Tested on STM32F072b and STM32F746

File: drivers/can/Kconfig.stm32 (hunk 4 above)
Comment (erwango) [reply drivers/can/Kconfig.stm32]: I meant 'tested on stm32F0 series'

File: drivers/can/can_stm32.c [hunk 6]
```diff
@@ -232,42 +276,36 @@ int can_stm32_runtime_configure(struct device *dev, enum can_mode mode,
...
-	hcan.Init.BS2  = bs2;
-	hcan.Init.Prescaler = prescaler;
-
-	hcan.State = HAL_CAN_STATE_RESET;
-
-	hal_ret = HAL_CAN_Init(&hcan);
-	if (hal_ret != HAL_OK) {
-		LOG_ERR("HAL_CAN_Init failed: %d", hal_ret);
-		return -EIO;
+	reg_mode =  mode == CAN_NORMAL_MODE   ? 0U   :
```
Comment (erwango) [inline drivers/can/can_stm32.c]: We need parenthesis here.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/can/Kconfig.stm32 (hunk 4 above)
Comment (alexanderwachter) [reply drivers/can/Kconfig.stm32]: This comment is from a very early stage of the driver where we only had support for this CPU.
I think i can safely remove it now.

**Review by erwango — APPROVED** [review]

**Review by galak — APPROVED** [review]
